nupic/research/frameworks/pytorch/datasets/permuted_mnist.py

```diff
@@ -0,0 +1,146 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import os
+
+import torch
+from torchvision import transforms
+from torchvision.datasets import MNIST
+
+from nupic.research.frameworks.dendrites.routing import generate_context_vectors
+
+
+class PermutedMNIST(MNIST):
+    """
+    The permutedMNIST dataset contains MNIST images where the same random permutation
+    of pixel values is applied to each image. More specifically, the dataset can be
+    broken down into 'tasks', where each such task is the set of all MNIST images, but
+    with a unique pixel-wise permutation applied to all images. `num_tasks` gives the
+    number of 10-way classification tasks (each corresponding to a unique pixel-wise
+    permutation) that a continual learner will try to learn in sequence.
+    """
+
+    # Permutations should be the same whether this class is instantiated for training
+    # or testing
+    permutations = None
+
+    def __init__(self, num_tasks, train, root=".", target_transform=None,
+                 download=False):
+
+        data_transform = transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize((0.13062755,), (0.30810780,)),
+        ])
+        super().__init__(root=root, train=train, transform=data_transform,
+                         target_transform=target_transform, download=download)
+
+        self.num_tasks = num_tasks
+
+        # Generate num_tasks random permutations - the first one is the identity
+        # permutation (i.e., regular MNIST), represented below as `None`
+        if PermutedMNIST.permutations is None:
+            PermutedMNIST.permutations = [
```

Comment:
`permutations` is clearly an instance variable, since it is modified in the init method of a class (the init initializes an instance). Why is it defined as a class variable?

Reply:
Since we instantiate a `PermutedMNIST` object twice (once for training and once for validation), we need the same pixel-wise permutations applied to images for each task. I found that defining `permutations` as a class variable is the easiest way to accomplish this.

One other way to accomplish the same would be to make `permutations` an optional argument to the `__init__` method, and that way when the validation class is instantiated, we can pass in the training class's permutations so that the same pixel-wise permutations are also applied to the validation examples. However, this may get messy since I'm not sure how the configs in Vernon would do this exactly.

Do you have any preferences on which way to go?

Reply:
There are several ways of solving this, but I don't think class attributes is one of them. Even though it might work in one tested used case, that is not what class attributes are for and can cause hidden bugs. 

Training and validation could share the dataloader. If they can't, and dataloader has to share permutations, then composition makes sense - permutations belonging to a higher level class (like Experiments), and passed as an object to dataloader. 

But the easiest way to solve is just seed it. Have a seed be initialized in the higher class and be passed as an argument to the dataloader, no need to pass all permutations. This will always yield the same results: 
```
g = torch.Generator()
g.manual_seed(30)
torch.randperm(100, generator=g)
```

Reply:
I think this is a good suggestion. I modified the dataset class so that `seed` is now an argument (specified in the config), and that's used to generate the permutations and contexts.